### ISSUE TITLE
TinyG G18 XZ Plane Seems to be Reversed

### ISSUE THREAD

**Issue by thrushd** [synthetos/TinyG#120]
Hello,

I recently upgraded my custom CNC to TinyG. So far I am absolutely loving the intuitive settings and ease of use, unfortunately a small issue popped up. When my machine goes to do an arc (using G2 and G3, but it also happens in radius mode) in the G18 plane it will move in the opposite direction it is intended (CW instead of CCW). It only seems to happen in the G18 plane. I noticed this when trying a simple facing operation for my vacuum bed, the specific G18 arcs are being used for the lead ins. They were very small though so it was hard to tell at first. This prompted me to make a quick test piece to experiment on, just a 50mm x 50mm block. 

![block_with_tool_path](https://cloud.githubusercontent.com/assets/8573804/5672158/117a859c-9743-11e4-94a5-3ef221b09873.png)

Did the same basic facing operation but this time made the lead in and out radii much bigger, 10mm. It is very noticeable now. Also loaded it into the HSMWorks backplotter and played with the code. Reversing the G2 and G3 shows what my machine is actually doing:

![hsmworks_backplot_mod](https://cloud.githubusercontent.com/assets/8573804/5672197/55dd61fa-9743-11e4-86f5-685d9aa599cd.png)

But when I use that modified reversed code on the machine it behaves great. So it seems to me that the G18 plane is reversed on the TinyG. I have played with all the settings I can think of, nothing changes. Used a variety of post processors and got the same results so I have been sticking to the built in "generic tinyg". Completely out of ideas and haven't been able to find anyone else with this issue. Any help or insight greatly appreciated!

**Details:**

CAD: Solidworks 2014 Education Edition
CAM: HSMWorks 2014

TinyG Settings Dump:

```
[fb]  firmware build            438.02
[fv]  firmware version            0.97
[hp]  hardware platform           1.00
[hv]  hardware version            8.00
[id]  TinyG ID                    3X3566-SRM
[ja]  junction acceleration  100000 mm
[ct]  chordal tolerance           0.0100 mm
[sl]  soft limit enable           0
[st]  switch type                 0 [0=NO,1=NC]
[mt]  motor idle timeout         30.00 Sec
[ej]  enable json mode            0 [0=text,1=JSON]
[jv]  json verbosity              1 [0=silent,1=footer,2=messages,3=configs,4=linenum,5=verbose]
[js]  json serialize style        1 [0=relaxed,1=strict]
[tv]  text verbosity              1 [0=silent,1=verbose]
[qv]  queue report verbosity      2 [0=off,1=single,2=triple]
[sv]  status report verbosity     1 [0=off,1=filtered,2=verbose]
[si]  status interval           250 ms
[ec]  expand LF to CRLF on TX     0 [0=off,1=on]
[ee]  enable echo                 0 [0=off,1=on]
[ex]  enable flow control         1 [0=off,1=XON/XOFF, 2=RTS/CTS]
[baud] USB baud rate              5 [1=9600,2=19200,3=38400,4=57600,5=115200,6=230400]
[net] network mode                0 [0=master]
[gpl] default gcode plane         0 [0=G17,1=G18,2=G19]
[gun] default gcode units mode    1 [0=G20,1=G21]
[gco] default gcode coord system  1 [1-6 (G54-G59)]
[gpa] default gcode path control  2 [0=G61,1=G61.1,2=G64]
[gdi] default gcode distance mode 0 [0=G90,1=G91]
[1ma] m1 map to axis              0 [0=X,1=Y,2=Z...]
[1sa] m1 step angle               1.800 deg
[1tr] m1 travel per revolution    5.0000 mm
[1mi] m1 microsteps               8 [1,2,4,8]
[1po] m1 polarity                 0 [0=normal,1=reverse]
[1pm] m1 power management         2 [0=disabled,1=always on,2=in cycle,3=when moving]
[2ma] m2 map to axis              1 [0=X,1=Y,2=Z...]
[2sa] m2 step angle               1.800 deg
[2tr] m2 travel per revolution    5.0000 mm
[2mi] m2 microsteps               8 [1,2,4,8]
[2po] m2 polarity                 0 [0=normal,1=reverse]
[2pm] m2 power management         2 [0=disabled,1=always on,2=in cycle,3=when moving]
[3ma] m3 map to axis              2 [0=X,1=Y,2=Z...]
[3sa] m3 step angle               1.800 deg
[3tr] m3 travel per revolution    5.0000 mm
[3mi] m3 microsteps               8 [1,2,4,8]
[3po] m3 polarity                 0 [0=normal,1=reverse]
[3pm] m3 power management         2 [0=disabled,1=always on,2=in cycle,3=when moving]
[4ma] m4 map to axis              3 [0=X,1=Y,2=Z...]
[4sa] m4 step angle               1.800 deg
[4tr] m4 travel per revolution  360.0000 mm
[4mi] m4 microsteps               8 [1,2,4,8]
[4po] m4 polarity                 0 [0=normal,1=reverse]
[4pm] m4 power management         2 [0=disabled,1=always on,2=in cycle,3=when moving]
[xam] x axis mode                 1 [standard]
[xvm] x velocity maximum       1200 mm/min
[xfr] x feedrate maximum       1200 mm/min
[xtn] x travel minimum        -1000.000 mm
[xtm] x travel maximum            0.000 mm
[xjm] x jerk maximum            100 mm/min^3 * 1 million
[xjh] x jerk homing           10000 mm/min^3 * 1 million
[xjd] x junction deviation        0.0500 mm (larger is faster)
[xsn] x switch min                0 [0=off,1=homing,2=limit,3=limit+homing]
[xsx] x switch max                1 [0=off,1=homing,2=limit,3=limit+homing]
[xsv] x search velocity         800 mm/min
[xlv] x latch velocity          100 mm/min
[xlb] x latch backoff             5.000 mm
[xzb] x zero backoff              3.000 mm
[yam] y axis mode                 1 [standard]
[yvm] y velocity maximum       1200 mm/min
[yfr] y feedrate maximum       1200 mm/min
[ytn] y travel minimum         -500.000 mm
[ytm] y travel maximum            0.000 mm
[yjm] y jerk maximum            100 mm/min^3 * 1 million
[yjh] y jerk homing           10000 mm/min^3 * 1 million
[yjd] y junction deviation        0.0500 mm (larger is faster)
[ysn] y switch min                0 [0=off,1=homing,2=limit,3=limit+homing]
[ysx] y switch max                1 [0=off,1=homing,2=limit,3=limit+homing]
[ysv] y search velocity         800 mm/min
[ylv] y latch velocity          100 mm/min
[ylb] y latch backoff             5.000 mm
[yzb] y zero backoff              3.000 mm
[zam] z axis mode                 1 [standard]
[zvm] z velocity maximum       1200 mm/min
[zfr] z feedrate maximum       1200 mm/min
[ztn] z travel minimum         -150.000 mm
[ztm] z travel maximum            0.000 mm
[zjm] z jerk maximum            100 mm/min^3 * 1 million
[zjh] z jerk homing           10000 mm/min^3 * 1 million
[zjd] z junction deviation        0.0500 mm (larger is faster)
[zsn] z switch min                0 [0=off,1=homing,2=limit,3=limit+homing]
[zsx] z switch max                1 [0=off,1=homing,2=limit,3=limit+homing]
[zsv] z search velocity         800 mm/min
[zlv] z latch velocity          100 mm/min
[zlb] z latch backoff             5.000 mm
[zzb] z zero backoff              3.000 mm
[aam] a axis mode                 3 [radius]
[avm] a velocity maximum     230400 deg/min
[afr] a feedrate maximum     230400 deg/min
[atn] a travel minimum           -1.000 deg
[atm] a travel maximum           -1.000 deg
[ajm] a jerk maximum           5760 deg/min^3 * 1 million
[ajh] a jerk homing            5760 deg/min^3 * 1 million
[ajd] a junction deviation        0.0500 deg (larger is faster)
[ara] a radius value              0.1989 deg
[asn] a switch min                1 [0=off,1=homing,2=limit,3=limit+homing]
[asx] a switch max                0 [0=off,1=homing,2=limit,3=limit+homing]
[asv] a search velocity         600 deg/min
[alv] a latch velocity          100 deg/min
[alb] a latch backoff             5.000 deg
[azb] a zero backoff              2.000 deg
[bam] b axis mode                 0 [disabled]
[bvm] b velocity maximum       3600 deg/min
[bfr] b feedrate maximum       3600 deg/min
[btn] b travel minimum           -1.000 deg
[btm] b travel maximum           -1.000 deg
[bjm] b jerk maximum             20 deg/min^3 * 1 million
[bjd] b junction deviation        0.0500 deg (larger is faster)
[bra] b radius value              1.0000 deg
[cam] c axis mode                 0 [disabled]
[cvm] c velocity maximum       3600 deg/min
[cfr] c feedrate maximum       3600 deg/min
[ctn] c travel minimum           -1.000 deg
[ctm] c travel maximum           -1.000 deg
[cjm] c jerk maximum             20 deg/min^3 * 1 million
[cjd] c junction deviation        0.0500 deg (larger is faster)
[cra] c radius value              1.0000 deg
[p1frq] pwm frequency            125000 Hz
[p1csl] pwm cw speed lo               0 RPM
[p1csh] pwm cw speed hi           11520 RPM
[p1cpl] pwm cw phase lo           0.000 [0..1]
[p1cph] pwm cw phase hi           1.000 [0..1]
[p1wsl] pwm ccw speed lo           1000 RPM
[p1wsh] pwm ccw speed hi           2000 RPM
[p1wpl] pwm ccw phase lo          0.125 [0..1]
[p1wph] pwm ccw phase hi          0.200 [0..1]
[p1pof] pwm phase off             0.000 [0..1]
[g54x] g54 x offset               0.000 mm
[g54y] g54 y offset               0.000 mm
[g54z] g54 z offset               0.000 mm
[g54a] g54 a offset               0.000 deg
[g54b] g54 b offset               0.000 deg
[g54c] g54 c offset               0.000 deg
[g55x] g55 x offset              75.000 mm
[g55y] g55 y offset              75.000 mm
[g55z] g55 z offset               0.000 mm
[g55a] g55 a offset               0.000 deg
[g55b] g55 b offset               0.000 deg
[g55c] g55 c offset               0.000 deg
[g56x] g56 x offset               0.000 mm
[g56y] g56 y offset               0.000 mm
[g56z] g56 z offset               0.000 mm
[g56a] g56 a offset               0.000 deg
[g56b] g56 b offset               0.000 deg
[g56c] g56 c offset               0.000 deg
[g57x] g57 x offset               0.000 mm
[g57y] g57 y offset               0.000 mm
[g57z] g57 z offset               0.000 mm
[g57a] g57 a offset               0.000 deg
[g57b] g57 b offset               0.000 deg
[g57c] g57 c offset               0.000 deg
[g58x] g58 x offset               0.000 mm
[g58y] g58 y offset               0.000 mm
[g58z] g58 z offset               0.000 mm
[g58a] g58 a offset               0.000 deg
[g58b] g58 b offset               0.000 deg
[g58c] g58 c offset               0.000 deg
[g59x] g59 x offset               0.000 mm
[g59y] g59 y offset               0.000 mm
[g59z] g59 z offset               0.000 mm
[g59a] g59 a offset               0.000 deg
[g59b] g59 b offset               0.000 deg
[g59c] g59 c offset               0.000 deg
[g92x] g92 x offset               0.000 mm
[g92y] g92 y offset               0.000 mm
[g92z] g92 z offset               0.000 mm
[g92a] g92 a offset               0.000 deg
[g92b] g92 b offset               0.000 deg
[g92c] g92 c offset               0.000 deg
[g28x] g28 x position             0.000 mm
[g28y] g28 y position             0.000 mm
[g28z] g28 z position             0.000 mm
[g28a] g28 a position             0.000 deg
[g28b] g28 b position             0.000 deg
[g28c] g28 c position             0.000 deg
[g30x] g30 x position             0.000 mm
[g30y] g30 y position             0.000 mm
[g30z] g30 z position             0.000 mm
[g30a] g30 a position             0.000 deg
[g30b] g30 b position             0.000 deg
[g30c] g30 c position             0.000 deg
tinyg [mm] ok> 
```

Thanks!

**Comment by aldenhart** [comment]
That's for the heads up and the detailed. This is a bug that should be corrected.

**Comment by thrushd** [comment]
So this is just a bug in the current firmware? Nothing more I can go until the next release? Thanks!

**Comment by aldenhart** [comment]
We'll try to get to this quickly and post an update for an experimental release if you are interested.

**Comment by thrushd** [comment]
I would definitely be interested, just keep me posted, thanks!

**Comment by aldenhart** [comment]
Hi. Checking back in after some work on the arcs - It looks like the G18 arcs are actually working correctly, unless I'm misinterpreting things. TinyG follows the NIST and Linux CNC definition of Gcode. From here:

http://linuxcnc.org/docs/html/gcode/gcode.html#sec:G2-G3-Arc 
"A circular or helical arc is specified using either G2 (clockwise arc) or G3 (counterclockwise arc) at the current feed rate. The direction (CW, CCW) is as viewed from the positive end of the axis about which the circular motion occurs."

So G18 G2... should trace a circle in the clockwise direction about Y as viewed from the rear of the machine (positive Y), or a counter-clockwise arc as viewed from the front of the machine. This is a move such as:
g21 g18 g2 i20 x40 f300    or
g21 g18 g2 i20 f300  (full circle)

I observe this to rotate in the CCW direction as viewed from the front. Is this your understanding? As I said, I could be misinterpreting things.

**Comment by aldenhart** [comment]
Please pardon any interim chatter. I have been able to reproduce your issue.

**Comment by aldenhart** [comment]
I have a fix in test. Should be able to push to edge in the next few days.

**Comment by aldenhart** [comment]
This issue is fixed in edge build 440.13 (and above). Please let me know if you have any questions.

**Comment by thrushd** [comment]
Awesome, thanks!
